### PR TITLE
Feature/real time saving results

### PR DESCRIPTION
**Pull request by jjmartinezQT**
This PR brings the possibility to use real time saving results capability.

### CODE REVIEW

File: src/qililab/result/stream_results.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    This class wraps a numpy array and adds a context manager to save results on real time while they are acquired by
+    the instruments.
+
+    Args:
+        shape (ndarray.shape): results array shape.
+        path (str): path to save results.
+        loops (dict[str, ndarray]): dictionary with each loop name in the experiment as key and numpy array as values.
+    """
+
+    def __init__(self, shape, path, loops):
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:17]: Don't forget to add types! 🙏

File: tests/result/test_stream_results.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    def test_stream_array_instantiation(self, stream_array: StreamArray):
+        """Tests the instantiation of a StreamArray object."""
+        assert stream_array.results.shape == (2, 2)
+        assert (stream_array.results == np.empty(shape=(2, 2))).all
+        assert stream_array.path == "test_stream_array.hdf5"
+        assert stream_array.loops == {"test_amp_loop": np.arange(0, 1, 2)}
+
+    @patch("qililab.result.stream_results.h5py.File", return_value=MockFile())
+    def test_context_manager(self, mock_h5py: MockFile, stream_array: StreamArray):  # pylint: disable=unused-argument
```
Comment (AlbertMitjans) [inline tests/result/test_stream_results.py:63]: We should use `mock_h5py` to assert that the right calls are being made

File: src/qililab/result/stream_results.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        self.results = np.empty(shape=shape)
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:20]: Let's use zeros instead:
image

File: src/qililab/result/stream_results.py (hunk 1 above)
Comment (jjmartinezQT) [reply src/qililab/result/stream_results.py:17]: added now.

File: src/qililab/result/stream_results.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        path (str): path to save results.
+        loops (dict[str, np.ndarray]): dictionary with each loop name in the experiment as key and numpy array as values.
+    """
+
+    def __init__(self, shape: tuple, path: str, loops: dict[str, np.ndarray]):
+        self.results = np.empty(shape=shape)
+        self.path = path
+        self.loops = loops
+        self.file: h5py.File | None = None
+        self.dataset = None
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:24]: Will users use these attributes? If not, let's make them private 🙏

Comment (AlbertMitjans) [reply src/qililab/result/stream_results.py:24]: I would keep `path` and `loops` public though, because these are the values they give in the init (this is only my humble opinion).

In other words, I expect to have access to the information I provided to a class. Here's an example:

```pycon
>>> a = A(hello="world")
>>> a.hello
world
```

File: src/qililab/result/stream_results.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
+from typing import Any
+
+import h5py
+import numpy as np
+
+
+class StreamArray:
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:7]: Finally, I want to ask for a change that is solely to improve the user experience.

This is something that I've been wanting to do for a long time now, and this is the perfect example to start doing it.

If you check popular python libraries (like numpy), you'll see that all of its classes/functions are shown in lowercase to the user, for example:
```pycon
>>> import numpy as np
>>> ar = np.array([1, 2])  # array is a class, but it is lowercase!
```

This is solely to improve user experience. The user shouldn't care if something is a class or not, thus they shouldn't care about lower/upper case letters.

In PennyLane, they solved this in a way that I really liked: they kept the syntax of upper-case letters for classes, but they created "constructors" (a.k.a functions) in lowercase that returned an instance of the class. In the PennyLane main `__init__` file, only these lower-case functions were imported (similar thing for documentation). [See an example in this link](https://github.com/PennyLaneAI/pennylane/blob/f25eb134d1aaca35c9cd886cd38f4283b457728e/pennylane/ops/op_math/pow.py#L39).

What do you think of doing the same thing? In this case it would be something like this:
```suggestion
def stream_results(shape: tuple, path: str, loops: dict[str, np.ndarray]):  # this is the constructor
    return StreamArray(shape=shape, path=path, loops=loops)

class StreamArray:
```

If we do this, the big chunk of the documentation (long explanation + examples) should be moved to this user-facing method, because this is what users will see. 😄 

Let me know if you have any questions!

**Review by AlbertMitjans — COMMENTED** [review]
Amazing job!! I added some comments 😄

File: src/qililab/result/stream_results.py (hunk 5 above)
Comment (jjmartinezQT) [reply src/qililab/result/stream_results.py:7]: I like the idea actually and it we enforce all along qililab little by little I think it will indeed improve user experience.

File: src/qililab/result/stream_results.py (hunk 3 above)
Comment (jjmartinezQT) [reply src/qililab/result/stream_results.py]: addressed.

Comment (jjmartinezQT) [reply src/qililab/result/stream_results.py]: addressed

File: src/qililab/result/stream_results.py (hunk 4 above)
Comment (jjmartinezQT) [reply src/qililab/result/stream_results.py]: agree with this change.

File: src/qililab/result/stream_results.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+def stream_results(shape: tuple, path: str, loops: dict[str, np.ndarray]):
+    """Constructs a StreamArray instance.
+
+    This methods serves as a constructor for user interface of the StreamArray class.
+
+    Args:
+        shape (tuple): results array shape.
+        path (str): path to save results.
+        loops (dict[str, np.ndarray]): dictionary with each loop name in the experiment as key and numpy array as values.
+    """
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:16]: Can we add an example? 🙏

**Review by AlbertMitjans — COMMENTED** [review]
Can we import the `stream_results` function in the top init, and add it into the documentation? 🙏 

We should also prepare a demo for next sprint review, such that HW knows how to use this feature!

File: tests/result/test_stream_results.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    def __init__(self):
+        """Initialize a mock file."""
+        self.dataset = None
+
+    def create_group(self, name: str):  # pylint: disable=unused-argument
+        return MockGroup()
+
+    def create_dataset(self, name: str, data: np.ndarray):  # pylint: disable=unused-argument
+        """Creates a dataset"""
+        return data
```
Comment (AlbertMitjans) [inline tests/result/test_stream_results.py:47]: You should return a `copy.deepcopy` of this!

File: src/qililab/result/stream_results.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        self._dataset = None
+
+    def __setitem__(self, key: str, value: float):
+        """Sets and item by key and value in the dataset.
+
+        Args:
+            key (str): key for the item to save.
+            value (float): value to save.
+        """
+        self._results[key] = value
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:117]: We should also modify the `self.dataset` attribute!

File: tests/result/test_stream_results.py (hunk 7 above)
Comment (AlbertMitjans) [reply tests/result/test_stream_results.py:47]: This is because h5py creates a copy of the data before putting it inside the dataset

File: docs/releases/changelog-dev.md [hunk 9]
```diff
@@ -2,6 +2,11 @@
 
 ### New features since last release
 
+- Added real time results saving capability.
+  [#598](https://github.com/qilimanjaro-tech/qililab/pull/598)
+
+- Added the two main classes need for automatic-calibration, `CalibrationController` and `CalibrationNode`: [#554](https://github.com/qilimanjaro-tech/qililab/pull/554)
```
Comment (AlbertMitjans) [inline docs/releases/changelog-dev.md:8]: This should not be here, right? 😆

File: src/qililab/result/stream_results.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    Args:
+        shape (tuple): results array shape.
+        path (str): path to save results.
+        loops (dict[str, np.ndarray]): dictionary with each loop name in the experiment as key and numpy array as values.
+
+    Examples:
+
+        .. note::
+
+            All the following examples are explained in detail in the :ref:`Platform <platform>` section of the documentation. However, here are a few thing to keep in mind:
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:34]: Juanjo did you copy this from somewhere else? 😅 

Let's try to keep the examples short and concise please! Otherwise nobody will read them 🙏 

I don't think you should be talking about the Platform here... you should only add an example of how this function works. If you want (and if needed), you can always add a **small** note that links to the platform documentation.

So please, remove everything in this example that is not related to the `stream_results` function 🙏

Comment (AlbertMitjans) [reply src/qililab/result/stream_results.py:34]: Also, can you add an extra example that explains how to use this with the `%%submit_job` magic method of SLURM?

**Review by AlbertMitjans — COMMENTED** [review]
Just some minor comments, but almost good to go!

File: src/qililab/drivers/instruments/qblox/test_cluster.ipynb
Comment (AlbertMitjans) [inline src/qililab/drivers/instruments/qblox/test_cluster.ipynb:1]: What is this file? Juanjo please make sure to check the PR changes before marking it ready for review!

File: src/qililab/result/stream_results.py [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+            # Connecting and setting up the platform:
+            platform.connect()
+            platform.initial_setup()
+            platform.turn_on_instruments()
+
+            # Define circuit
+            circuit = Circuit(QUBIT + 1)
+            circuit.add(ql.Drag(QUBIT, theta=np.pi, phase=0))
+            circuit.add(ql.Wait(QUBIT, M_BUFFER_TIME))
+            circuit.add(gates.M(QUBIT))
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:63]: Again, this example is way too long. It contains a lot of information that is not related to this method's functionality.

All the lines I marked above can be deleted. A good example could be:

```python
import qililab as ql

stream_array = ql.stream_results(shape=(5, 2), loops={"amp": AMP_VALUES}, path="results.h5")

with stream_array:
    for i in range(5):
        stream_array[i, 0] = i
```
```pycon
>>> stream_array
# Show output
>>> ql.load_results("results.h5")
# Show output
```

Comment (AlbertMitjans) [reply src/qililab/result/stream_results.py:63]: The example above is super short, copy-pasteable, and concise (it shows the functionality of the feature).

Comment (AlbertMitjans) [reply src/qililab/result/stream_results.py:63]: Then, if we want to add more information (for example, related to the slurm magic method) we can add an extra example below:

Comment (AlbertMitjans) [reply src/qililab/result/stream_results.py:63]: For example:

> .. note::
> 
>     The output of `stream_results` is not picklable, thus it cannot be returned in a SLURM job. In this scenario we suggest copying the array with the data into a different variable and returning that variable instead:
> 
>     .. code-block:: python
> 
>         %%submit_job -o results -d galadriel
>         stream = ql.stream_results(shape=(5, 2), loops={"amp": AMP_VALUES}, path="results.h5")
>         with stream:
>             for i in range(5):
>                 # Here you can execute any algorithm you want
>                 stream[i] = i
>         results = stream.results

File: src/qililab/result/stream_results.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        loops (dict[str, np.ndarray]): dictionary with each loop name in the experiment as key and numpy array as values.
+
+    Examples:
+
+        **1. Executing an experiment and saving the results live time:**
+
+
+        To execute an experiment you first need to define the loops and the values you want to loop for, for example, a Rabi experiment
+        where you will loop over amplitude values, measuring the results out of the amplitude in each of the steps.
+        Then you also need to build, connect, set up, and execute the platform, which together look like:
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:37]: This text is completely unrelated to the code below

File: src/qililab/result/stream_results.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+            [3. 0.]
+            [4. 0.]]
+
+            >>> ql.load_results("results.h5")
+
+            (array([[0., 0.],
+                [1., 0.],
+                [2., 0.],
+                [3., 0.],
+                [4., 0.]]), {'amp': array([0.  , 0.25, 0.5 , 0.75, 1.  ])})
```
[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply src/qililab/result/stream_results.py:65]: We might also need to remove the empty space between the command and the output

File: src/qililab/result/stream_results.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+         [0.5  0.  ]
+         [0.75 0.  ]
+         [1.   0.  ]]
+
+        >>> ql.load_results("results.h5")
+        (array([[0., 0.],
+               [0.25, 0.],
+               [0.5, 0.],
+               [0.75, 0.],
+               [1., 0.]]), {'amp': array([0.  , 0.25, 0.5 , 0.75, 1.  ])})
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:62]:
```suggestion
               [1., 0.]]), {'loop_name': array([0.  , 0.25, 0.5 , 0.75, 1.  ])})
```

File: src/qililab/result/stream_results.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+    .. note::
+
+        The output of `stream_results` is not picklable, thus it cannot be returned in a SLURM job. In this scenario we suggest copying the array with the data into a different variable and returning that variable instead:
+
+        .. code-block:: python
+
+            %%submit_job -o results -d galadriel
+            LOOP_VALUES = np.linspace(0, 1, 5)
+            stream_array = ql.stream_results(shape=(5, 2), loops={"loop_name": LOOP_VALUES}, path="results.h5")
+            with stream:
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:73]:
```suggestion
            with stream_array:
```

File: src/qililab/result/stream_results.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+                for i, value in enumerate(LOOP_VALUES):
+                    # Here you can execute any algorithm you want
+                    stream_array[(i, 0)] = value
+
+            results = stream.results
```
Comment (AlbertMitjans) [inline src/qililab/result/stream_results.py:78]:
```suggestion
            results = stream_array.results
```

**Review by AlbertMitjans — APPROVED** [review]
Good job!! 🚀 

I left some comments but it is ready to go!
